x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/action/CO2Calculator.java

```diff
@@ -43,33 +47,79 @@ final class CO2Calculator {
             : customPerCoreWattARM64 / 1000.0d;
     }
 
+    /**
+     * Calculates the annual CO2 emissions in metric tons for a given host and sample count.
+     *
+     * @param hostID The ID of the host
+     * @param samples The number of samples
+     * @return The annual CO2 emissions in metric tons
+     */
     public double getAnnualCO2Tons(String hostID, long samples) {
+        // Calculate annual core hours based on sampling duration and frequency
         double annualCoreHours = CostCalculator.annualCoreHours(samplingDurationInSeconds, samples, DEFAULT_SAMPLING_FREQUENCY);
 
+        // Retrieve host metadata for the given hostID
         HostMetadata host = hostMetadata.get(hostID);
         if (host == null) {
+            logger.warning("Host metadata is null for hostID: " + hostID);
```

Comment:
Missing host metadata is expected and may be quite common. Logging this is would be too verbose.